### PR TITLE
Connect JSON editor to modal

### PR DESCRIPTION
**Pull request by MikeTschudi**
* Valid changes in the JSON editor
  * activate the Cancel & Save modal buttons
  * are snapshotted to the underlying component store
* Invalid changes in the JSON
  * are flagged as an error in the editor
  * disable the Cancel & Save modal buttons regardless of the state of changes in the solution
  * are not snapshotted

Snapshots are kept even as one switches among solution items.

### CODE REVIEW

File: src/components/json-editor/json-editor.tsx [hunk 1]
```diff
@@ -387,6 +401,56 @@ export class JsonEditor {
     document.getElementById(buttonId)?.removeAttribute("disabled");
   }
 
+  /**
+   * Dispatches an event that the editor's content has changed.
+   */
+  protected _flagEditorContentChanged(): void {
+    // Event for notifying that the editor contents have changed
+    window.dispatchEvent(new CustomEvent("solutionEditorContentChanged", {
```
Comment (jmhauck) [inline src/components/json-editor/json-editor.tsx:409]: @MikeTschudi does dispatchEvent allow us to do something that stencils EventEmitter doesn't?

File: src/components/json-editor/json-editor.tsx [hunk 2]
```diff
@@ -387,6 +401,56 @@ export class JsonEditor {
...
+      }));
+      this.hasChanges = flagHasChanges;
+    }
+  }
+
+  /**
+   * Sets the editor's flag indicating if it has errors and dispatches an event when
+   * the flag value changes.
+   */
+  protected _flagEditorHasErrors(flagHasErrors: boolean): void {
```
Comment (jmhauck) [inline src/components/json-editor/json-editor.tsx:441]: doc here as well

File: src/components/json-editor/json-editor.tsx [hunk 3]
```diff
@@ -387,6 +401,56 @@ export class JsonEditor {
...
+      cancelable: false,
+      composed: true
+    }));
+  }
+
+  /**
+   * Sets the editor's flag indicating if it has changes and dispatches an event when
+   * the flag value changes.
+   */
+  protected _flagEditorHasChanges(flagHasChanges: boolean): void {
```
Comment (jmhauck) [inline src/components/json-editor/json-editor.tsx:424]: could you doc the arg please

File: src/components/solution-configuration/solution-configuration.tsx [hunk 4]
```diff
@@ -76,6 +76,27 @@ export class SolutionConfiguration {
   //
   //--------------------------------------------------------------------------
 
+  constructor() {
```
Comment (jmhauck) [inline src/components/solution-configuration/solution-configuration.tsx:79]: @MikeTschudi similar thing here with addEventListener vs stencils @ Listen...is this done for a particular reason?

File: src/components/solution-configuration/solution-configuration.tsx [hunk 5]
```diff
@@ -333,6 +359,25 @@ export class SolutionConfiguration {
     this._treeOpen = !this._treeOpen;
   }
 
+  protected _updateSaveability(
```
Comment (jmhauck) [inline src/components/solution-configuration/solution-configuration.tsx:362]: doc?

File: src/components/solution-template-data/solution-template-data.tsx [hunk 6]
```diff
@@ -90,6 +91,29 @@ export class SolutionTemplateData {
   //
   //--------------------------------------------------------------------------
 
+  constructor() {
+    window.addEventListener("solutionEditorContentChanged",
```
Comment (jmhauck) [inline src/components/solution-template-data/solution-template-data.tsx:95]: same thing here with addEventListener vs Listen

File: src/demos/new-solution-configuration.html [hunk 7]
```diff
@@ -199,10 +199,10 @@ <h1>Demo Solution Configuration</h1>
           modalTitle.innerHTML = `Configure Solution ${itemBase.title}`;
 
           configurationComponent.solutionItemId = solutionId;
-          window.addEventListener("solutionStoreHasChanges",
+          window.addEventListener("solutionCanSave",
```
Comment (jmhauck) [inline src/demos/new-solution-configuration.html:202]: addEventListener vs Listen

File: src/utils/solution-store.ts [hunk 8]
```diff
@@ -347,18 +347,18 @@ class SolutionStore
    * Sets the store's flag indicating if it has changes and dispatches an event when
    * the flag value changes.
    */
-  protected _flagStoreAsChanged(flagAsChanged: boolean): void {
+  protected _flagStoreHasChanges(flagHasChanges: boolean): void {
```
Comment (jmhauck) [inline src/utils/solution-store.ts:350]: doc arg

**Review by jmhauck — CHANGES_REQUESTED** [review]

File: src/components/json-editor/json-editor.tsx (hunk 1 above)
Comment (MikeTschudi) [reply src/components/json-editor/json-editor.tsx:409]: It does less in that StencilJS handles deallocating listeners for us with the @Event/@Listener markers, but I haven't been able to get them to work

**Review by jmhauck — APPROVED** [review]

**Pull request merged**